File: P5 2024-2/Paradigmas de Programacao/09-06/Python/Animal.py
class Animal:
    def __init__(self, especie, nome):
        self.especie = especie
        self.nome = nome
    
    def emitir_som(self):
        pass

class Cachorro(Animal):
    def emitir_som(self):
        return "Au Au"
    
class Gato(Animal):
    def emitir_som(self):
        return "Miau"
    
class Mamifero(Animal):
    def __init__(self, especie, nome):
        super().__init__(especie, nome)

    def amamentar(self):
        print(f"{self.nome} está amamentando")

class Ave(Animal):
    def __init__(self, especie, nome):
        super().__init__(especie, nome)

    def voar(self):
        print(f"{self.nome} está voando")

class Morcego(Mamifero, Ave):
    def __init__(self, especie, nome):
        super().__init__(especie, nome)

    def emitir_som(self):
        return "Eu sou o batman"

File: P5 2024-2/Paradigmas de Programacao/09-06/Golang/Animal.go
package main

import (
	"fmt"
)

// Definindo a estrutura Triangulo
type Animal struct {
	especie string
	nome    string
}

func (a Animal) emitirSom() {
	fmt.Println(a.nome, "faz um barulho...")
}

type Cachorro struct {
	Animal
	especie string
}

func (c Cachorro) emitirSom() {
	fmt.Println("Au Au")
}

type Gato struct {
	Animal
	especie string
}

func (g Gato) emitirSom() {
	fmt.Println("Miau")
}

type Mamifero struct {
	Animal
}

func (m Mamifero) amamentar() {
	fmt.Println(m.nome, "está amamentando...")
}

type Ave struct {
	Animal
}

func (a Ave) voar() {
	fmt.Println(a.nome, "está voando...")
}

type Morcego struct {
	Ave
	Mamifero
}
